File: napier/napier.c
#include<stdio.h>
#include<stdlib.h>
#include<time.h>

int trial() {
    int counter = 0;
    double sum = 0;
    while (1) {
        sum += (double)rand() / (double)RAND_MAX;
        counter += 1;
        if ( sum >= 1.0)
            break;
    }
    return counter;
}

int main() {
    srand(time(NULL));
    int n_trial = 100000000;
    int sum = 0;
    for (int i = 0; i < n_trial; i++) {
        sum += trial();
    }
    printf("E(N) = %f\n", (double)sum / (double)n_trial);
    return 0;
}

File: napier/napier.go
package main

import (
    "fmt"
    "math/rand"
    "time"
)

func count_upto_one() int {
    s := rand.NewSource(time.Now().UnixNano())
    r := rand.New(s)
    counter := 0
    accumurated := 0.0
    for true {
        accumurated += r.Float64()
        counter += 1
        if accumurated > 1.0 {
            break
        }
    }
    return counter
}

func main() {
    fmt.Println("Start")
    n_trial := int(1e8)
    total_count := 0
    for i := 0; i < n_trial; i++ {
        total_count += count_upto_one()
    }
    fmt.Println("napier=", float64(total_count)/float64(n_trial))
}
